File: server/lib/src/main/base/db/schema.dart
import 'package:json_annotation/json_annotation.dart';

import '../../utils/version.dart';
import 'annotations/schema.dart';

part 'schema.g.dart';

enum Privileges {
  read,
  write,
  publish,
  ultimate;

  static Privileges fromString(String name) {
    return Privileges.values
        .singleWhere((v) => v.toString() == name.toLowerCase());
  }
}

enum VCS {
  git,
  svn,
  fossil,
  mercurial,
  other;

  static VCS fromString(String name) {
    return VCS.values.singleWhere((v) => v.toString() == name.toLowerCase());
  }
}

/// A package, as represented in the pritt database
class Package {
  /// The id of the package
  @primary
  String id;

  /// The name of the package
  @unique
  String name;

  /// The description of the package
  String? description;

  /// The latest version of the given package
  /// updated with every publish to the database
  String version;

  /// The person who published the given package
  ///
  /// This is a foreign reference to a user
  User author;

  /// The programming language
  String language;

  /// The last published time of the given package
  DateTime updated;

  /// The first published time of the given package
  DateTime created;

  /// The VCS that the package uses for development
  ///
  /// This information is not used at the moment, and is reserved for future iterations
  VCS vcs;

  /// The VCS URL of the package, if applicable
  Uri? vcsUrl;

  /// The archive directory path of the given package.
  ///
  /// This archive is usually for the Object File System and so is relative to that
  Uri archive;

  /// The license of the given package
  String? license;

  /// Whether a package is scoped or not
  bool get scoped => scope != null;

  /// The scope of the package, if it is scoped
  String? scope;

  Package(
      {required this.id,
      required this.name,
      required this.version,
      required this.author,
      required this.language,
      this.description,
      DateTime? updated,
      required this.created,
      this.vcs = VCS.git,
      required this.archive,
      this.license,
      this.scope,
      this.vcsUrl})
      : updated = updated ?? created;
}

/// Maps packages to their versions, and info about those versions
class PackageVersions {
  @primary
  @ForeignKey(Package, property: 'name')
  Package package;

  String version;

  /// The type of version here
  VersionType versionType;

  /// The published time of the given package
  DateTime created;

  /// The contents of the readme file associated with the package
  String? readme;

  /// The raw contents of the config file associated with the config
  String? config;

  /// The name of the configuration file used for the given package
  String? configName;

  /// Some info about this package, like the size and more
  Map<String, dynamic> info;

  /// Environment information (runtime, package manager versions, etc)
  /// e.g npm, node
  Map<String, String> env;

  /// Metadata about the package
  ///
  /// This varies between programming languages based on schema
  /// e.g npmUser
  Map<String, dynamic> metadata;

  /// The archive path of the given package.
  ///
  /// This archive is usually for the Object File System and so is relative to that
  Uri archive;

  /// The archive SHA256 hash data
  String hash;

  /// Signatures for the given package when published
  List<Signature> signatures;

  /// The integrity hash of the given package
  String integrity;

  /// Whether the given package is deprecated
  bool isDeprecated;

  /// A deprecation message for the given package
  String? deprecationMessage;

  /// Whether a given package is yanked
  bool isYanked;

  PackageVersions(
      {required this.package,
      required this.version,
      required this.versionType,
      DateTime? updated,
      required this.created,
      this.readme,
      this.config,
      this.configName,
      required this.info,
      required this.env,
      required this.metadata,
      required this.archive,
      required this.hash,
      required this.signatures,
      required this.integrity,
      this.isDeprecated = false,
      this.isYanked = false,
      this.deprecationMessage})
      : assert(config == null || configName != null,
            "If config is set, then configName must be set as well");
}

/// Join table for contributors for a package
class PackageContributors {
  @ForeignKey(Package, property: 'id')
  Package package;

  @ForeignKey(User, property: 'id')
  User contributor;

  /// When the contributor was added to the package
  DateTime addedAt;

  /// The kind of privileges this contributor has, when contributing to this package.
  ///
  /// No contributor can have [Privileges.ultimate] except the author himself, unless he passes the package down to another person (not possible in pritt yet)
  ///
  /// Users can have a combination of privileges.
  Iterable<Privileges> privileges;

  PackageContributors({
    required this.package,
    required this.contributor,
    required this.privileges,
    required this.addedAt,
  }) : assert(
            privileges.contains(Privileges.ultimate) && privileges.length == 1,
            "Ultimate privilege cannot be combined with read privilege");
}

/// User information
///
/// TODO: Auth?
class User {
  /// The id of the user
  @primary
  String id;

  /// The name of the user
  String name;

  /// The email address of the user
  String email;

  /// The time the user joined
  DateTime createdAt;

  /// The last time the user updated information
  DateTime updatedAt;

  User(
      {required this.id,
      required this.name,
      required this.email,
      required this.createdAt,
      required this.updatedAt});
}

// class NewUser extends User {
//   /// The current access token for the given user
//   ///
//   /// This is used for authenticating workflows for the CLI, installing packages, etc
//   // String accessToken;
// }

/// A scope is an organizational unit for packages
@Table('organizations')
class Scope {
  /// The id of the scope
  @primary
  String id;

  /// The name of the scope
  @unique
  String name;

  /// The description of the scope
  String? description;

  /// The time the scope was created
  DateTime createdAt;

  /// The time the scope was last updated
  DateTime updatedAt;

  Scope({
    required this.id,
    required this.name,
    this.description,
    required this.createdAt,
    required this.updatedAt,
  });
}

class AccessToken {
  String id;

  /// The user
  String userId;

  /// The hashed version of the access token
  String hash;

  /// The type of access token used
  /// Could be a personal one [AccessTokenType.personal], or for the CLI [AccessTokenType.device], or for Actions [AccessTokenType.pipeline]
  AccessTokenType tokenType;

  String? description;

  /// The device id, if any
  String? deviceId;

  /// When the access token expires
  DateTime expiresAt;

  /// When the access token was last used
  DateTime lastUsedAt;

  /// When the access token was created
  DateTime createdAt;

  /// The device information
  Map<String, dynamic>? deviceInfo;

  AccessToken(
      {required this.id,
      required this.userId,
      required this.hash,
      required this.tokenType,
      this.description,
      this.deviceId,
      required this.expiresAt,
      required this.lastUsedAt,
      required this.createdAt,
      this.deviceInfo});
}

enum AccessTokenType {
  device,
  personal,
  extended,
  pipeline;

  static AccessTokenType fromString(String name) => AccessTokenType.values.firstWhere((v) => v.name == name);
}

/// A join table for users and scopes
@Table('organization_members')
class ScopeUsers {
  @ForeignKey(Scope, property: 'id')
  @Key(name: 'organization_id')
  Scope scope;

  @ForeignKey(User, property: 'id')
  @Key(name: 'user_id')
  User user;

  /// The privileges the user has in the scope
  Iterable<Privileges> privileges;

  /// When the user was added to the scope
  DateTime joinedAt;

  ScopeUsers({
    required this.scope,
    required this.user,
    required this.privileges,
    required this.joinedAt,
  });
}

@JsonSerializable()
class Signature {
  /// The public key id of the signature
  String publicKeyId;

  /// The signature itself
  String signature;

  /// The time the signature was created
  DateTime created;

  Signature({
    required this.publicKeyId,
    required this.signature,
    required this.created,
  });

  factory Signature.fromJson(Map<String, dynamic> json) =>
      _$SignatureFromJson(json);
  Map<String, dynamic> toJson() => _$SignatureToJson(this);
}

class Plugin {
  @primary
  String id;

  String name;

  String language;

  String? description;

  Uri archive;

  PluginArchiveType archiveType;

  Plugin({
    required this.id,
    required this.name,
    required this.language,
    this.description,
    required this.archive,
    this.archiveType = PluginArchiveType.single,
  });
}

enum PluginArchiveType {
  single,
  multi;

  static PluginArchiveType fromString(String name) =>
      PluginArchiveType.values.firstWhere((v) => v.name == name);
}

/// Authorization Session Tasks
class AuthorizationSession {
  String id;
  @Key(name: 'session_id')
  String sessionId;

  String? userId;

  TaskStatus status;

  @Key(name: 'expires_at')
  DateTime expiresAt;

  /// The device ID information
  String deviceId;

  String? accessToken;

  /// When the auth flow started
  DateTime startedAt;

  /// When the auth flow was authorized
  DateTime? authorizedAt;

  /// A code to authorize with
  String code;

  AuthorizationSession({
    required this.id,
    required this.sessionId,
    this.userId,
    required this.deviceId,
    this.status = TaskStatus.pending,
    required this.expiresAt,
    required this.startedAt,
    this.authorizedAt,
    this.accessToken,
    required this.code,
  }) {
    if (expiresAt.isBefore(DateTime.now()) && status == TaskStatus.pending) {
      status = TaskStatus.expired;
    }
  }
}

enum TaskStatus {
  pending,
  success,
  fail,
  expired,
  error;

  static TaskStatus fromString(String name) =>
      TaskStatus.values.firstWhere((v) => v.name == name);
}

File: web/server/db/schema/schema.ts
import { pgTable, index, integer, varchar, timestamp, boolean, unique, check, text, foreignKey, uuid, jsonb, json, primaryKey, pgEnum } from "drizzle-orm/pg-core"
import { sql } from "drizzle-orm"

export const accessTokenType = pgEnum("access_token_type", ['device', 'personal', 'extended', 'pipeline'])
export const pluginArchiveType = pgEnum("plugin_archive_type", ['single', 'multi'])
export const pluginSourceType = pgEnum("plugin_source_type", ['hosted', 'vcs', 'local', 'other'])
export const privilege = pgEnum("privilege", ['read', 'write', 'publish', 'ultimate'])
export const taskStatus = pgEnum("task_status", ['pending', 'success', 'fail', 'expired', 'error', 'idle', 'queue'])
export const versionControlSystem = pgEnum("version_control_system", ['git', 'svn', 'fossil', 'mercurial', 'other'])
export const versionKind = pgEnum("version_kind", ['major', 'experimental', 'beta', 'next', 'rc', 'canary', 'other'])


export const flywaySchemaHistory = pgTable("flyway_schema_history", {
	installedRank: integer("installed_rank").primaryKey().notNull(),
	version: varchar({ length: 50 }),
	description: varchar({ length: 200 }).notNull(),
	type: varchar({ length: 20 }).notNull(),
	script: varchar({ length: 1000 }).notNull(),
	checksum: integer(),
	installedBy: varchar("installed_by", { length: 100 }).notNull(),
	installedOn: timestamp("installed_on", { mode: 'string' }).defaultNow().notNull(),
	executionTime: integer("execution_time").notNull(),
	success: boolean().notNull(),
}, (table) => [
	index("flyway_schema_history_s_idx").using("btree", table.success.asc().nullsLast().op("bool_ops")),
]);

export const plugins = pgTable("plugins", {
	id: text().primaryKey().notNull(),
	name: text().notNull(),
	language: text().notNull(),
	description: text(),
	archive: text().notNull(),
	archiveType: pluginArchiveType("archive_type").default('single').notNull(),
	sourceType: pluginSourceType("source_type").default('other').notNull(),
	url: text(),
	vcs: versionControlSystem(),
}, (table) => [
	unique("plugins_name_key").on(table.name),
	unique("plugins_language_key").on(table.language),
	check("valid_name", sql`name ~ '^[a-zA-Z0-9][a-zA-Z0-9_.-]*$'::text`),
	check("url_source_type_check", sql`(source_type <> ALL (ARRAY['vcs'::plugin_source_type, 'hosted'::plugin_source_type])) OR (url IS NOT NULL)`),
	check("vcs_source_type_check", sql`(source_type <> 'vcs'::plugin_source_type) OR (vcs IS NOT NULL)`),
]);

export const authorizationSessions = pgTable("authorization_sessions", {
	id: uuid().default(sql`uuid_generate_v4()`).primaryKey().notNull(),
	sessionId: text("session_id").notNull(),
	userId: text("user_id"),
	status: taskStatus().default('pending').notNull(),
	authorizedAt: timestamp("authorized_at", { withTimezone: true, mode: 'string' }),
	startedAt: timestamp("started_at", { withTimezone: true, mode: 'string' }).defaultNow().notNull(),
	expiresAt: timestamp("expires_at", { withTimezone: true, mode: 'string' }).notNull(),
	deviceId: text("device_id").notNull(),
	code: varchar().notNull(),
	accessToken: text("access_token"),
}, (table) => [
	index("idx_login_sessions_expiry").using("btree", table.expiresAt.asc().nullsLast().op("timestamptz_ops")),
	foreignKey({
			columns: [table.userId],
			foreignColumns: [users.id],
			name: "authorization_sessions_user_id_fkey"
		}).onDelete("set null"),
	unique("authorization_sessions_session_id_key").on(table.sessionId),
	unique("authorization_sessions_code_key").on(table.code),
]);

export const packagePublishingTasks = pgTable("package_publishing_tasks", {
	id: uuid().default(sql`uuid_generate_v4()`).primaryKey().notNull(),
	status: taskStatus().default('queue').notNull(),
	userId: text("user_id").notNull(),
	name: text().notNull(),
	scope: text(),
	version: text().notNull(),
	new: boolean().notNull(),
	language: text().notNull(),
	config: text().notNull(),
	configMap: jsonb("config_map").notNull(),
	metadata: jsonb().default({}).notNull(),
	env: json().default({}).notNull(),
	vcs: versionControlSystem().default('other').notNull(),
	vcsUrl: text("vcs_url"),
	updatedAt: timestamp("updated_at", { withTimezone: true, mode: 'string' }).defaultNow().notNull(),
	createdAt: timestamp("created_at", { withTimezone: true, mode: 'string' }).defaultNow().notNull(),
	expiresAt: timestamp("expires_at", { withTimezone: true, mode: 'string' }).notNull(),
}, (table) => [
	foreignKey({
			columns: [table.userId],
			foreignColumns: [users.id],
			name: "package_publishing_tasks_user_id_fkey"
		}).onDelete("set null"),
]);

export const organizations = pgTable("organizations", {
	id: text().primaryKey().notNull(),
	name: text().notNull(),
	description: text(),
	createdAt: timestamp("created_at", { withTimezone: true, mode: 'string' }).defaultNow().notNull(),
	updatedAt: timestamp("updated_at", { withTimezone: true, mode: 'string' }).defaultNow().notNull(),
}, (table) => [
	unique("organizations_name_key").on(table.name),
]);

export const users = pgTable("users", {
	id: text().primaryKey().notNull(),
	name: text().notNull(),
	email: text().notNull(),
	avatarUrl: text("avatar_url").notNull(),
	createdAt: timestamp("created_at", { withTimezone: true, mode: 'string' }).defaultNow().notNull(),
	updatedAt: timestamp("updated_at", { withTimezone: true, mode: 'string' }).defaultNow().notNull(),
});

export const accessTokens = pgTable("access_tokens", {
	id: uuid().default(sql`uuid_generate_v4()`).primaryKey().notNull(),
	userId: text("user_id").notNull(),
	hash: text().notNull(),
	tokenType: accessTokenType("token_type").notNull(),
	description: text(),
	deviceId: text("device_id"),
	expiresAt: timestamp("expires_at", { withTimezone: true, mode: 'string' }).notNull(),
	lastUsedAt: timestamp("last_used_at", { withTimezone: true, mode: 'string' }).defaultNow().notNull(),
	createdAt: timestamp("created_at", { withTimezone: true, mode: 'string' }).defaultNow(),
	deviceInfo: jsonb("device_info"),
}, (table) => [
	foreignKey({
			columns: [table.userId],
			foreignColumns: [users.id],
			name: "access_tokens_user_id_fkey"
		}).onDelete("cascade"),
	unique("access_tokens_hash_key").on(table.hash),
]);

export const packages = pgTable("packages", {
	id: text().primaryKey().notNull(),
	name: text().notNull(),
	version: text().notNull(),
	description: text(),
	authorId: text("author_id").notNull(),
	scope: text(),
	language: text().notNull(),
	createdAt: timestamp("created_at", { withTimezone: true, mode: 'string' }).defaultNow().notNull(),
	updatedAt: timestamp("updated_at", { withTimezone: true, mode: 'string' }).defaultNow().notNull(),
	vcs: versionControlSystem().default('git').notNull(),
	vcsUrl: text("vcs_url"),
	archive: text().notNull(),
	license: text(),
}, (table) => [
	foreignKey({
			columns: [table.authorId],
			foreignColumns: [users.id],
			name: "packages_author_id_fkey"
		}),
	foreignKey({
			columns: [table.scope],
			foreignColumns: [organizations.name],
			name: "packages_scope_fkey"
		}),
	unique("packages_name_scope_key").on(table.name, table.scope),
	unique("packages_version_key").on(table.version),
	check("valid_name", sql`name ~ '^[a-zA-Z0-9][a-zA-Z0-9_.-]*$'::text`),
	check("valid_scope", sql`scope ~ '^[a-zA-Z0-9][a-zA-Z0-9_.-]*$'::text`),
]);

export const organizationMembers = pgTable("organization_members", {
	organizationId: text("organization_id").notNull(),
	userId: text("user_id").notNull(),
	privileges: privilege().array().notNull(),
	joinedAt: timestamp("joined_at", { withTimezone: true, mode: 'string' }).defaultNow().notNull(),
}, (table) => [
	foreignKey({
			columns: [table.organizationId],
			foreignColumns: [organizations.id],
			name: "organization_members_organization_id_fkey"
		}),
	foreignKey({
			columns: [table.userId],
			foreignColumns: [users.id],
			name: "organization_members_user_id_fkey"
		}),
	primaryKey({ columns: [table.organizationId, table.userId], name: "organization_members_pkey"}),
]);

export const packageContributors = pgTable("package_contributors", {
	packageId: text("package_id").notNull(),
	contributorId: text("contributor_id").notNull(),
	privileges: privilege().array().notNull(),
	joinedAt: timestamp("joined_at", { withTimezone: true, mode: 'string' }).defaultNow().notNull(),
}, (table) => [
	foreignKey({
			columns: [table.packageId],
			foreignColumns: [packages.id],
			name: "package_contributors_package_id_fkey"
		}),
	foreignKey({
			columns: [table.contributorId],
			foreignColumns: [users.id],
			name: "package_contributors_contributor_id_fkey"
		}),
	primaryKey({ columns: [table.packageId, table.contributorId], name: "package_contributors_pkey"}),
]);

export const packageVersions = pgTable("package_versions", {
	packageId: text("package_id").notNull(),
	version: text().notNull(),
	versionType: versionKind("version_type").default('major').notNull(),
	createdAt: timestamp("created_at", { withTimezone: true, mode: 'string' }).defaultNow().notNull(),
	readme: text(),
	config: text(),
	configName: text("config_name"),
	info: jsonb().default({}).notNull(),
	env: json().default({}).notNull(),
	metadata: jsonb().default({}).notNull(),
	archive: text().notNull(),
	hash: text().notNull(),
	signatures: jsonb().notNull(),
	integrity: text().notNull(),
	deprecated: boolean().default(false).notNull(),
	deprecatedMessage: text("deprecated_message"),
	yanked: boolean().default(false).notNull(),
}, (table) => [
	foreignKey({
			columns: [table.packageId],
			foreignColumns: [packages.id],
			name: "package_versions_package_id_fkey"
		}),
	primaryKey({ columns: [table.packageId, table.version], name: "package_versions_pkey"}),
]);
